python/paddle/fluid/tests/unittests/test_count_nonzero_api.py

```diff
@@ -0,0 +1,112 @@
+from paddle.fluid import Program, program_guard
+
+np.random.seed(10)
+
+
+class TestCountNonzeroAPI(unittest.TestCase):
+    # test paddle.tensor.math.count_nonzero
+
+    def setUp(self):
+        self.x_shape = [2, 3, 4, 5]
+        self.x = np.random.uniform(-1, 1, self.x_shape).astype(np.float32)
+        self.place = paddle.CUDAPlace(0) if core.is_compiled_with_cuda() \
+            else paddle.CPUPlace()
+
+    def test_api_static(self):
+        paddle.enable_static()
+        with paddle.static.program_guard(paddle.static.Program()):
+            x = paddle.fluid.data('X', self.x_shape)
+            out1 = paddle.count_nonzero(x)
+            out2 = paddle.tensor.count_nonzero(x)
+            out3 = paddle.tensor.math.count_nonzero(x)
+            axis = np.arange(len(self.x_shape)).tolist()
+            out4 = paddle.count_nonzero(x, axis)
+            out5 = paddle.count_nonzero(x, tuple(axis))
+            exe = paddle.static.Executor(self.place)
+            res = exe.run(feed={'X': self.x},
+                          fetch_list=[out1, out2, out3, out4, out5])
+        out_ref = np.count_nonzero(self.x)
+        for out in res:
+            self.assertEqual(np.allclose(out, out_ref), True)
+
+    def test_api_dygraph(self):
+        paddle.disable_static(self.place)
+
+        def test_case(x, axis=None, keepdim=False):
+            x_tensor = paddle.to_tensor(x)
+            out = paddle.count_nonzero(x_tensor, axis=axis, keepdim=keepdim)
+            if isinstance(axis, list):
+                axis = tuple(axis)
+                if len(axis) == 0:
+                    axis = None
+
+            out_ref = np.count_nonzero(x, axis, keepdims=keepdim)
+            self.assertEqual(np.allclose(out.numpy(), out_ref), True)
+
+        test_case(self.x)
+        test_case(self.x, None)
+        test_case(self.x, -1)
+        test_case(self.x, keepdim=True)
+        test_case(self.x, 2, keepdim=True)
+        test_case(self.x, [0, 2])
+        test_case(self.x, (0, 2))
+        test_case(self.x, (0, 1, 3))
+        test_case(self.x, [0, 1, 2, 3])
+        paddle.enable_static()
+
+    def test_errors(self):
+        paddle.enable_static()
+        with paddle.static.program_guard(paddle.static.Program()):
+            x = paddle.fluid.data('X', [10, 12], 'int32')
+            self.assertRaises(ValueError, paddle.count_nonzero, x, axis=10)
+
+    def test_api_dygraph_grad(self):
+        paddle.disable_static(self.place)
+
+        def test_case(x, axis=None, keepdim=False):
+            if isinstance(axis, list):
+                axis = list(axis)
+                if len(axis) == 0:
+                    axis = None
+            x_tensor = paddle.to_tensor(x, stop_gradient=False)
+            y = paddle.count_nonzero(x_tensor, axis, keepdim)
+            dx = paddle.grad(y, x_tensor)[0].numpy()
+            self.assertEqual(
+                np.allclose(np.sum(np.ones_like(x, dtype=np.int64)), dx.sum()),
+                True)
```

Comment:
Is the gradient of x obtained by this code correct in math? or x should have gradient?

Reply:
Thanks for reminding me, Mr. Gao. Here this count operator was mistaken for sum operator by me. According to gradient law, x is not supposed to have gradient. @jeff41404